index/internal/collections/collections.go

```
@@ -0,0 +1,121 @@
+// Copyright 2025 EngFlow Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collections
+
+import (
+	"maps"
+	"slices"
+)
+
+func Map[T, V any](ts []T, fn func(T) V) []V {
+	result := make([]V, len(ts))
+	for i, t := range ts {
+		result[i] = fn(t)
+	}
+	return result
+}
+
+func FlatMap[T, V any](ts []T, fn func(T) []V) []V {
+	result := []V{}
+	for _, t := range ts {
+		result = slices.AppendSeq(result, slices.Values(fn(t)))
+	}
+	return result
+}
+
+func Collect[T, V any](ts []T, fn func(T) (V, error)) []V {
```

I don't like that this hides the errors. What will we use this for? I imagine it would be better to have something like:

```
func FilterMap[T, V any](ts []T, fn func(T) (V, bool)) []V
```

where the result from `fn` is included in the returned slice if the second return value is `true`.

---

It was mostly used to collect entries that might fail for some, here's one example from conan integration: 

```go 
target := &indexer.ModuleTarget{
	Name: name,
	Hdrs: collections.ToSet(collections.Collect(
		info.GetNamedAttribute("hdrs").GetStringListValue(),
		label.Parse)),
	Includes:           collections.ToSet(info.GetNamedAttribute("includes").GetStringListValue()),
	StripIncludePrefix: info.GetNamedAttribute("strip_include_prefix").GetStringValue(),
	IncludePrefix:      info.GetNamedAttribute("include_prefix").GetStringValue(),
	Deps: collections.ToSet(collections.Collect(
		info.GetNamedAttribute("deps").GetStringListValue(),
		label.Parse)),
}
``` 
Here `label.Parse` is a function that tries to parse string as bazel label, but can fail returning error. We don't care about entries that failed this parsing, eg. empty strings, malformed inputs. 

If we 'd go with FilterMap it would look more or less like: 
```go
collections.FilterMap(
    info.GetNamedAttribute("deps").GetStringListValue(),
    func(s string) (label.Label, bool) {
        v, err := label.Parse(s)
        return v, err == nil
})
```

---

I think I like that `FilterMap` usage better, since it makes it more explicit what we're doing with the error. We could extract that lambda into a local helper function if we need to use it more than once.